### ISSUE TITLE
[v2 FEATURE]: Increase customizability of activation functions

### ISSUE THREAD

**Issue by shihchengli** [chemprop/chemprop#878]
**Is your feature request related to a problem? Please describe.**
Currently, only the activation functions defined in nn/utils.py can be used, which limits the customizability of activation functions.

**Discussion**
We can provide the flexibility to accept both `Activation` and `nn.Module` objects as inputs for activation functions (floowing the format like [this](https://pytorch.org/docs/stable/_modules/torch/nn/modules/activation.html#ReLU)). Additionally, we can add an activation attribute to the neural network, similar to what PyTorch Geometric does ([link](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.MLP.html#torch_geometric.nn.models.MLP)). I am open to any suggestions.

**Comment by davidegraff** [comment]
I don't like the design pattern specifying both the initializer function (or an alias) _as well as_ the parameters as inputs into a separate class. At that point, you're using composition without actually designing around it, so it's an all-around loss. For example, compare the following two design patterns using the `Bar` class:

```python
class Bar(ABC):
    def __init__(*args, **kwargs):
        ...

class BarA(Bar): ...
class BarB(Bar): ...

def get_bar(bar_type: str) -> type[Bar]: ...
```

1. Implicit composition:

```
class Foo:
    def __init__(self, bar_type: str, bar_args, bar_kwargs):
       self.bar = get_bar(bar_type)(*bar_args, **bar_kwargs)
```

2. Explicit Composition

```
class Foo:
    def __init__(self, bar):
       self.bar = bar
```

I think (2) is significantly more clear of what's going on, as it doesn't require users to cross reference any functions or initializers to see what's going on. All of this is to say, that I'm fine with moving away from the usage of `ActivationType` inside core module code and favoring explicitly passing `activation: nn.Module`. We can move `get_activation_function` into the CLI, as there's really no need to restrict the activation types used inside our core modules. This was a reimplementation of v1 logic, but I don't agree with it (even if I chose to write it).

**Comment by shihchengli** [comment]
Thanks for sharing your thoughts. I agree explicitly passing the activation function into the MessagePassing and Predictor blocks is a clear way to improve both customizability and code readability. Do you disagree with this idea because you think passing a callable function into a neural network builder as an attribute is not good practice?

**Comment by KnathanM** [comment]
Closed by #1146
